beacon-chain/p2p/gossipsub_peer_crawler.go

```diff
@@ -0,0 +1,515 @@
+			topics: make(map[gossipsubcrawler.Topic]struct{}),
+		}
+		cp.byEnode[enodeID] = existingPNode
+		cp.byPeerId[peerID] = existingPNode
+	} else {
+		existingPNode.node = node
+	}
+
+	cp.updateTopicsUnlocked(existingPNode, topics)
+
+	if existingPNode.isPinged || len(topics) == 0 {
+		cp.mu.Unlock()
+		return
+	}
+	cp.mu.Unlock()
+
+	select {
+	case cp.g.pingCh <- *node:
+	case <-cp.g.ctx.Done():
+		return
+	}
+}
+
+func (cp *crawledPeers) removeTopic(topic gossipsubcrawler.Topic) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	// Get all peers subscribed to this topic
+	peers, ok := cp.byTopic[topic]
+	if !ok {
+		return // Topic doesn't exist
+	}
+
+	// Remove the topic from each peer's topic list
+	for peerID := range peers {
+		if pnode, exists := cp.byPeerId[peerID]; exists {
+			delete(pnode.topics, topic)
+			// remove the peer if it has no more topics left
+			if len(pnode.topics) == 0 {
+				cp.updateTopicsUnlocked(pnode, nil)
+			}
+		}
+	}
+
+	// Remove the topic from byTopic map
+	delete(cp.byTopic, topic)
+}
+
+func (cp *crawledPeers) removePeerId(peerID peer.ID) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	pnode, exists := cp.byPeerId[peerID]
+	if !exists {
+		return
+	}
+
+	// Use updateTopicsUnlocked with empty topics to remove the peer
+	cp.updateTopicsUnlocked(pnode, nil)
+}
+
+func (cp *crawledPeers) removePeer(enodeID enode.ID) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+	pnode, exists := cp.byEnode[enodeID]
+	if !exists {
+		return
+	}
+	cp.updateTopicsUnlocked(pnode, nil)
+}
+
+// setting topics to empty will remove the peer completely.
+func (cp *crawledPeers) updateTopicsUnlocked(pnode *peerNode, topics []string) {
+	// If topics is empty, remove the peer completely.
+	if len(topics) == 0 {
+		delete(cp.byPeerId, pnode.peerID)
+		delete(cp.byEnode, pnode.id)
+		for t := range pnode.topics {
+			if peers, ok := cp.byTopic[t]; ok {
+				delete(peers, pnode.peerID)
+				if len(peers) == 0 {
+					delete(cp.byTopic, t)
+				}
+			}
+		}
+		pnode.topics = nil // Clear topics to indicate removal.
+		return
+	}
+
+	newTopics := make(map[gossipsubcrawler.Topic]struct{})
+	for _, t := range topics {
+		newTopics[gossipsubcrawler.Topic(t)] = struct{}{}
+	}
+
+	// Remove old topics that are no longer present.
+	for oldTopic := range pnode.topics {
+		if _, exists := newTopics[oldTopic]; !exists {
+			if peers, ok := cp.byTopic[oldTopic]; ok {
+				delete(peers, pnode.peerID)
+				if len(peers) == 0 {
+					delete(cp.byTopic, oldTopic)
+				}
+			}
+		}
+	}
+
+	// Add new topics.
+	for newTopic := range newTopics {
+		if _, exists := pnode.topics[newTopic]; !exists {
+			if _, ok := cp.byTopic[newTopic]; !ok {
+				cp.byTopic[newTopic] = make(map[peer.ID]struct{})
+			}
+			cp.byTopic[newTopic][pnode.peerID] = struct{}{}
+		}
+	}
+	pnode.topics = newTopics
+}
+
+type GossipsubPeerCrawler struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	crawlInterval time.Duration
+	crawlTimeout  time.Duration
+
+	crawledPeers *crawledPeers
```

Comment:
Seems there is some circular references here:
`GossipsubPeerCrawler` has a `*crawledPeers` field, and `crawledPeers` has a `*GossipsubPeerCrawler` field.